specs/vm/instruction_set.md

```diff
@@ -1386,6 +1387,53 @@ Cease VM execution and revert script effects. After a revert:
 1. All [OutputContract](../protocol/tx_format.md#outputcontract) outputs will have the same `balanceRoot` and `stateRoot` as on initialization.
 1. All [OutputVariable](../protocol/tx_format.md#outputvariable) outputs will have `to`, `amount`, and `asset_id` of zero.
 
+### SMO: Send message to output
+
+|             |                                                                                     |
+|-------------|-------------------------------------------------------------------------------------|
+| Description | Send a message to recipient address `MEM[$rA, 32]` with call abi `MEM[$rA + 32, $rB]` and `$rD` coins, with output `$rC`. |
+| Operation   | ```outputmessage(MEM[$fp, 32], MEM[$rA, 32], MEM[$rA + 32, $rB], $rD, $rC);```      |
+| Syntax      | `smo $rA, $rB, $rC, $rD`                                                            |
+| Encoding    | `0x00 rA rB rC rD`                                                                  |
+| Notes       |                                                                                     |
+
+Given helper `balanceOfStart(asset_id: byte[32]) -> uint32` which returns the memory address of the remaining free balance of `asset_id`, or panics if `asset_id` has no free balance remaining.
+
+Panic if:
+
+- `$rA + 32` overflows
+- `$rA + $rB + 32` overflows
+- `$rA + 32 > VM_MAX_RAM`
+- `$rA + $rB + 32 > VM_MAX_RAM`
+- `$rC > tx.outputsCount`
+- In an external context, if `$rD > MEM[balanceOfStart(0), 8]`
+- In an internal context, if `$rD` is greater than the balance of asset ID 0 of output with contract ID `MEM[$fp, 32]`
+- `tx.outputs[$rC].type != OutputType.Message`
+- `tx.outputs[$rC].recipient != 0 || tx.outputs[$rC].sender != 0`
+
+Append a receipt to the list of receipts, modifying `tx.receiptsRoot`:
```

Comment:
I think this receipt needs to be in the JSON ABI spec as well: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/abi.md#receipt

Reply:
Isn't this implied? I don't see any reference to the abi for any other opcode that creates receipts. Should we create a separate issue to add reference to the abi for all opcodes that mention receipts?

Reply:
If you look at the above link, you'll see each of the receipt types have a JSON definition. This is unrelated to the instruction set, hence why other instructions don't talk about what the JSON receipts look like.

Reply:
As for whether instructions that emit receipts should also link to the relevant sections in the JSON ABI spec...probably yes!